### PR TITLE
Add support for monthly recurrences 

### PR DESCRIPTION
**Pull request by Stroemgren**
Trying to add monthly support following the pull request from @bruteforcecat

### CODE REVIEW

**Comment by bducharme** [comment]
@Stroemgren @doughsay Is there any update on this? 

Also, did you consider adding the support for the `ICalendar` parser?

**Comment by Stroemgren** [comment]
I’m just hoping for a merge here. 

If there’s anything you need @doughsay you should let me know.

**Comment by doughsay** [comment]
Hey guys, sorry for the lack of movement on this stuff.  Monthly recurrences are "hard", and I'm just really hesitant about adding something to the library that doesn't fully conform to the iCalendar spec.

For example, here is what this PR does for monthly recurrence using date time `~N[2017-01-31 09:00:00]`:

```elixir
iex(5)> schedule = ~N[2017-01-31 09:00:00] |> Cocktail.schedule() |> Schedule.add_recurrence_rule(:monthly)
#Cocktail.Schedule<Monthly>
iex(6)> Cocktail.Schedule.to_i_calendar(schedule)
"DTSTART:20170131T090000\nRRULE:FREQ=MONTHLY"
iex(7)> Cocktail.Schedule.occurrences(schedule) |> Enum.take(10)
[~N[2017-01-31 09:00:00], ~N[2017-02-28 09:00:00], ~N[2017-03-31 09:00:00],
 ~N[2017-04-30 09:00:00], ~N[2017-05-31 09:00:00], ~N[2017-06-30 09:00:00],
 ~N[2017-07-31 09:00:00], ~N[2017-08-31 09:00:00], ~N[2017-09-30 09:00:00],
 ~N[2017-10-31 09:00:00]]
iex(8)>
```

And here is what my other library ([excal](https://github.com/peek-travel/excal)) does with the same rule and start date:

```elixir
iex(3)> {:ok, stream} = Excal.Recurrence.Stream.new("FREQ=MONTHLY", ~N[2017-01-31 09:00:00])
{:ok, #Function<55.33009823/2 in Stream.resource/3>}
iex(4)> Enum.take(stream, 10)
[~N[2017-01-31 09:00:00], ~N[2017-03-31 09:00:00], ~N[2017-05-31 09:00:00],
 ~N[2017-07-31 09:00:00], ~N[2017-08-31 09:00:00], ~N[2017-10-31 09:00:00],
 ~N[2017-12-31 09:00:00], ~N[2018-01-31 09:00:00], ~N[2018-03-31 09:00:00],
 ~N[2018-05-31 09:00:00]]
```

If we actually wanted "the last day of the month", then we would do it like this:

```elixir
iex(5)> {:ok, stream} = Excal.Recurrence.Stream.new("FREQ=MONTHLY;BYMONTHDAY=-1", ~N[2017-01-31 09:00:00])
{:ok, #Function<55.33009823/2 in Stream.resource/3>}
iex(6)> Enum.take(stream, 10)
[~N[2017-01-31 09:00:00], ~N[2017-02-28 09:00:00], ~N[2017-03-31 09:00:00],
 ~N[2017-04-30 09:00:00], ~N[2017-05-31 09:00:00], ~N[2017-06-30 09:00:00],
 ~N[2017-07-31 09:00:00], ~N[2017-08-31 09:00:00], ~N[2017-09-30 09:00:00],
 ~N[2017-10-31 09:00:00]]
```

But cocktail (using this PR) can't handle this case:

```elixir
Cocktail.Schedule.from_i_calendar("DTSTART:20170131T090000\nRRULE:FREQ=MONTHLY;BYMONTHDAY=-1")
{:error, {:invalid_frequency, 1}}
```

excal implements a `NIF` that uses the C reference implementation of the iCalendar spec directly, so it conforms to the spec as best as the C library does. It's been quite a while since I last delved into this stuff, so I can't remember the exact reasons for why it behaves this way, but I'm sure the details are berried somewhere in here: https://tools.ietf.org/html/rfc5545#section-3.3.10

Basically, what it comes down to is: conforming to the actual iCalendar spec is really hard, and that's why I stopped trying with cocktail and wrote excal instead.  That's not to say that I wouldn't like cocktail to implement a pure elixir version of the spec! It just seems pretty hard...

If you want to tackle fixing the above case, and try a few other complicated edge cases (test with excal, or an online icalendar recurrence rule tester), then I will gladly accept this PR. You can also see a bunch of example recurrence rules in the spec here: https://tools.ietf.org/html/rfc5545#section-3.8.5.3

**Comment by doughsay** [comment]
Sorry I spoke a bit too soon, it does look like you're supporting negative day of month, it's just not parsing correctly in `from_i_calendar`.  But there appears to be a bug in the negative case, see here:

```elixir
iex(10)> ~N[2017-01-01 06:00:00] |> Cocktail.schedule() |> Schedule.add_recurrence_rule(:monthly, days_of_month: [-1]) |> Cocktail.Schedule.occurrences() |> Enum.take(10)
[~N[2017-01-30 06:00:00], ~N[2017-02-27 06:00:00], ~N[2017-03-30 06:00:00],
 ~N[2017-04-29 06:00:00], ~N[2017-05-30 06:00:00], ~N[2017-06-29 06:00:00],
 ~N[2017-07-30 06:00:00], ~N[2017-08-30 06:00:00], ~N[2017-09-29 06:00:00],
 ~N[2017-10-30 06:00:00]]
```

Looks like it's just off by one...

So!  If we can fix the bug mentioned in my first comment regarding just a plain `monthly` including dates it shouldn't, fix the negative counting issue, and make it so `from_i_calendar` can parse the rrules with negative numbers in it, we might be a lot closer to getting this merged!

**Comment by peaceful-james** [comment]
@doughsay I made a PR to try and deal with the issues you raised:
https://github.com/Stroemgren/cocktail/pull/1

It's a bit naive but if you have time would you mind giving it some 👁️ ?

**Comment by doughsay** [comment]
superseded by https://github.com/peek-travel/cocktail/pull/138
